### ISSUE TITLE
Proposal: Rename OrderedCollection to List, Dictionary to Map

### ISSUE THREAD

**Issue by melkyades** [powerlang/powerlang#6]
Module system should help with compatibility (letting refs to Dictionary and OrderedCollection continue to work)

**Comment by gbracha** [comment]
Well, FWIW, those are the names we use in Newspeak. And for us, it's easy to rename things for local use, as every module has its own namespace. But YMMV.

**Comment by janvrany** [comment]
@gbracha yes, in some of our internal talks I expressed a wish to have something not unlike NewSpeak modules and imports, unifying the concept of module (package) / class / nested class / shared pool. I hope we'll have time to get to this "soonish"...

**Comment by gbracha** [comment]
Thanks. Let me know if you ever wish to discuss.

**Comment by melkyades** [comment]
> Well, FWIW, those are the names we use in Newspeak. And for us, it's easy to rename things for local use, as every module has its own namespace. But YMMV.

I didn't know that, so I'm glad I'm not alone.

> Thanks. Let me know if you ever wish to discuss.

For sure, it would be great to talk about this. My caveat about Jan's suggestion is that classes already have too many responsibilities: creating instances and defining their behavior; so adding more to them may cause more problems than what it would solve. What I'd like is to let them define namespaces, but indirectly (through some ivar?). I'd also like to think about dividing them so that shape gets independent from behavior.

**Comment by melkyades** [comment]
The other caveat is that the DMR introduces some restrictions: when things get more late bound, they may cause an infinite recursion on the DMR that wouldn't occur on a classic VM

**Comment by gbracha** [comment]
So classes in Newspeak still have the same responsibilities: creating instances and defining behavior (though you could argue that it is mixins that define behavior). It's just that instances now cover everything - scoping is now covered by method lookup, so things are simpler. And because everything now goes via a message passing interface, behavior is  independent of representation (shape). But these are generalities, and in any case I'm not here to convince you to change Smalltalk.

The real tradeoffs are different. You pay for the encapsulation  Newspeak gives you. You can't just grab everything from a global namespace. And once you add access control, you also find that you can't always get at what you want that easily.
Whitebox testing is a good example. Extension methods are another (assuming you think they are a good thing).

In short, the security/modularity discipline does sacrifice some convenience. The benefits come at some cost. Just as mirrors are a bit less convenient than accessing classes and method tables directly, but enable many things that are just don't work without them.

You can mitigate these costs during development, but that takes investment.  For example, we had to change the class browser UI model, because of nesting.

I would like to hear more about your point about late binding being a problem on the DMR.  In general, I'm wondering how much work it would be to define a Newspeak DMR using powerlang.

**Comment by melkyades** [comment]
> creating instances and defining behavior (though you could argue that it is mixins that define behavior)

actually in Bee the behavior is defined indirectly too (each class has its own Behavior, which points to the method dictionary and the next Behavior -of the superclass-). One of the questions I often ask myself is, could we have shape separated from the class hierarchy? would it be nice or ugly? (maybe in newspeak it's like that?) specially considering a desire of having everything late bound. But yes, I guess this would be a quite different system from Smalltalk. The "incremental improvement" I've talked about with Jan is to at least have all dictionary-based vars late bound: globals, pools and class vars. Basically, if some identifier is not an instance variable, then it is late bound. Maybe in the end it leads to "ok, why where ivars different? let's make them late bound too".

> Extension methods are another (assuming you think they are a good thing).
This is another whole topic. To me, it's not whether they are good or bad, but whether there's a nice way to live without them or have some thing as powerful with so little effort.

> I would like to hear more about your point about late binding being a problem on the DMR. In general, I'm wondering how much work it would be to define a Newspeak DMR using powerlang.

Having a system defined in itself leads to some chicken-and-egg problems that have to be shortcircuited. For example, in the DMR, the lookup methods are precomputed, and nativized with a special linker so that message sends in this closure can be emitted as invokes (as lookup methods form a small closure, you can know who will be the receiver of each send and know the method to be executed). Now, if you make name binding lazy, it is very likely that your name lookup will also require this same form of shortcircuiting. It seems totally doable, but it's also likely that it might require some more work than it looks in the first place (I'm worried/interested in the interaction bewteen shortcircuiting method lookup and shortcircuiting name lookup). And then regarding the Newspeak & DMR, I see quite some options, all of which include AOTing a small kernel and a JIT. Easier ones could be Bee kernel & bee jit, or bee kernel & cog jit (and on top of any of those load Newspeak); or in the "turtles-all-the-way-down", Newspeak kernel & Newspeak JIT.

**Comment by shingarov** [comment]
> in the DMR, the lookup methods are precomputed, and nativized with a special linker so that message sends in this closure can be emitted as invokes (as lookup methods form a small closure, you can know who will be the receiver of each send and know the method to be executed). Now, if you make name binding lazy, it is very likely that your name lookup will also require this same form of shortcircuiting.

To my ears, this sounds like screaming "JSR292" with its set of interesting situations (what if an exception occurs during custom lookup).  Shijie Xu at UNB recently published a dissertation on optimization of Method Handles in a J9 implementation.  It would be interesting to compare his approach to DMR.

**Comment by gbracha** [comment]
So name lookup and method lookup in Newspeak are the same thing in theory. In practice, access to temporaries is not really implemented as a lookup.  If one writes the implementation in Bee, I don't think much changes then. If one goes for the turtles-all-the-way-down solution (which is my goal in the end), it seems to me that the difference in Newspeak would be that any instance variable access in the lookup methods would have to be treated as you would treat a method lookup. This might also include access to enclosing objects. I would perhaps do this in phases, first in Bee and the  in Newspeak;
but I'd have to understand a lot more about the process before deciding.

**Comment by shingarov** [comment]
As it was suggested: to conserve the related Slack thread in BA-ST in a somewhat-more-permanent medium, I am cut-and-pasting it here:

@leandro:
> Acabo de crear el canal #smalltalk-ide-as para ver si podemos identificar buenas ideas para una IDE de Smalltalk.

@shingarov:

My favorite would be, "How to get rid of IDEs altogether."

Seriously, in my search for departure away from Smalltalk-80 towards the true Smalltalk, IDEs and "the" Browser are nearly crowning my "list of anti-Smalltalk concepts in Smalltalk-80".  This is for several reasons.

First: ossification.

In his obituary of Larry Tesler, Kent Beck relates the following story:

> The early Smalltalkers knew they needed a browser, a way to navigate, understand, and modify the code in the system. The core group went off to talk about it. Larry got pissed, just wrote a damn browser (the design of which, much to his disgust, persists today 45 years later), and showed it to them when they returned.

I had various discussions with fellow Smalltalkers about what the moral of this story is, and heard things along the lines of, "just do it, instead of sitting in committees";  I think such interpretation would be too shallow.  The key to understanding Larry here, is the indefinite article: "_a browser_" -- just one out of thousands of possibilities to be explored by the programmer.    But Smalltalk-80 fixated Larry's sketch into the status of "_The Browser_", thus reinforcing the barrier between "the application" and "the Smalltalk system".  Such ossification is exactly the opposite of what Smalltalk is all about.

The word "IDE" stinks even worse, suggesting imagery of "immutable, Smalltalk-vendor-provided tools."  (Ok, replace "vendor" with "implementer" and it doesn't improve much).

What we need instead, is to create such tools on the spot, at the moment the need arises.  For example, to design and implement the JIT-postmortem debug UI I showed at the hackathon, took 1 day.  Well, 2 days if we count in the time I spent familiarizing with Tudor's APIs.  Speaking of which -- I very much agree with the GT approach.  (In the Museum of History of Science in Cambridge, there is a spectacular collection of _visual models_ -- enough to convince anyone that the scientific "discovery" is at least as much about _seeing_ and _showing_ as it is about, well, _discovery_ proper).  At the same time I like to emphasize that interesting "[micro/tele/...]scopes" are only possible thanks to the interesting structures they are visualizing.  For example, Tudor's Debugger-as-Context-Inspector is only possible because we have Contexts in the first place.  This is nontrivial because Smalltalk still has to learn a lot from languages like ML which offers the programmer incredible depth of debugging insight -- through a primitive command-line interface.

This is a two-way street.

The ossification of the tools, and the ossification of the fundamental data structures, have been -- I think -- mutually-causing in Smalltalk-80.  And vice-versa: whatever interesting development of one, will inevitably cause interesting development of the other.  Today, I am altogether not happy about what constructions I can express in Smalltalk.  We have a straightforward mechanism for products (just stick the factors into the object's "slots").  For co-products... we are supposed to use class polymorphism.  Actually, we are supposed to express all constructions as combinations of "slots" and "subclassing".  This can be really awkward: think of `true` and `false` -- the simplest construction of choice between two discrete points -- how we have to jump through hoops to e.g. disallow creating additional instances of `True`, etc.  We have a whole bunch of these -- they are called Design Patterns -- solutions to problems that shouldn't be problems in the first place.

So, what I want in the first place is an algebra of universal constructions in my language.  And then I want to be able to play with them without crashing the live Smalltalk system -- think of trying to change Smalltalk's metaobject protocol, even though it is written in Smalltalk (for example, change Squeak's class-as-species-membership to Bee's class-as-a-behavior).  Finally, once I have a myriad of object models, each will inspire its own set of tools, will imply its own universe of possible IDEs.

@guilleamaral:

> speaking about ossification, it seems like not only you are questioning the skin but also the bones... I think your channel should have been "How to get rid of the whole thing!" then!
seriously, though we agree on that the path taken by GT seems to break with the tooling dogma, you see more structural changes on the way to a purer smalltalk... or would we name it differently?

@shingarov:

> or would we name it differently?

We would name it Smalltalk.

Let me provide some background of how this "Smalltalk of the future" project started.

At ESUG'14 in Cambridge, right after my presentation, we sat down at one of those little coffee-tables and had a chat with Dan Ingalls, and I asked: "why is Smalltalk doing so poorly these decades?" -- and he said: "Because we have no Smalltalk; this what we have is Smalltalk-80 and it's not a Smalltalk."

So.

To secure a future for Smalltalk, the first thing we need is to build one.  Good examples to follow would be Smalltalk-71, Smalltalk-72, Smalltalk-76.  I call mine "Smalltalk-25", but let me emphasize that I mean it to be "a Smalltalk" just like Larry sketched his browser to be "a browser".

> it seems like not only you are questioning the _skin_ but also the _bones_

Yes, absolutely.
To keep close to the present topic of IDEs:
If I am building a new-generation tool, and some core structure is blocking me, then I have no other choice than to redesign that structure.  (As a familiar illustration: imagine you are designing the Debugger, and the core runtime doesn't have Contexts.  At this point, you have only two options: (1) give up altogether, or (2) redesign the VM to allow reification of sends).  So, it's not that I *chose* to question-the-bones by *licence* (because-I-felt-like-it or something) -- but because I am forced to.

These forced redesign steps lead one to the next, forming chains of inevitability.  What's scary is how far this chain goes.

I started from the modest scope of designing a better debugger [for my own VM debugging needs] which would be relevant/competitive in the force of what the competition has in ~2020.  The very first thing I bumped into was that I was blocked by this VM fundamentally rooted in the Deutsch--Schiffman'83 design.  So you start reevaluating DS'83 from first principles, and you immediately discover that your life is defined by the larget context of the computing-technology stack.  Which means you are at a roadfork of two options.

One is, what I call "the SKAMP approach" after Alan Knight's article "The Smalltalk Kick-Ass MegaProcessor" (in Smalltalk Report ca. 1993).  The argument was that, dynamic languages such as Smalltalk are such a tiny niche that any benefit from a hardware-accelerated Smalltalk processor [and other layers in the technology stack] is meaningless given the economy of scale.  So as dynamic-language-VM designers, we are stuck to play by the rules given by the big guys.

This is all the more unfortunate because the layers of the technology stack are not mutually-neutral.  E.g. the processor and the language compiler are co-designed in a tight whole of cause-and-effect considerations.  To be outside of that whole, means that conforming to the rules becomes increasingly more difficult as the technology progresses forward, stretching this first "SKAMP approach" to its limit.

Two big tide-waves in the last decade have effected why modern dynamic-language-VM design has started to follow an opposite approach.  1:  We have witnessed the stretch actually break for [certain] production customers.  2: Dynamic languages are no longer niche: people buy computers (be that phones, servers, whatever) to run JS, Python, Java etc.

For example, a number of VM people (Mario, Jecel, myself,... many others) have formed the "J-Extension" Task Group within RISC-V (which is itself a "fifth generation" of SOAR (Smalltalk-on-a-RISC)), so the new ISA is co-designed together with the VM.

And now that we are playing in the big guys' sandbox, now this chain-of-inevitability extends even further.  Now we are embedded in the co-design of the technology stack (and collaborating on the Processor Architecture, the Verification, the Platform, etc.)  Further, the technology stack itself is embedded in the co-design of society.  Which itself is moving; for example, the values that existed in "the XEROX PARC decade" (referring to my message in #books here) may not correspond with today's dichotomy between the group of creators vs strict-consumers.

**Comment by leandro** [comment]
@shingarov I guess you mentioned the wrong Leandro in your comment and I came here just to inform you that.
